docs/blog - Static Site Security (#12893)
gatsbyjs/gatsby: docs/blog/2019-03-27-security-for-modern-web-frameworks/index.md

- **Network** - This kind of attack happens when an attacker can access to the network being used to access your site. It could be a rouge third-party in your cloud provider, or someone sitting across the coffee shop on the same public WiFi as your user. The exact attack methods vary, but are collectively referred to as a [Man-in-the-Middle](https://en.wikipedia.org/wiki/Man-in-the-middle_attack) attack.
  The defense here is HTTPS (TLS encryption). Other related technologies such as [Cross-Origin-Resource-Sharing](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS), [Strict-Transport-Security](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security), [Public Key Pinning](*https://developer.mozilla.org/en-US/docs/Web/HTTP/Public_Key_Pinning), and [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) all assist in securing network traffic. No matter who is looking at your network traffic, they can not read or modify it.

> Other possible attacks relate to using `iframes` or not removing referrer tags on external links. Thankfully, fixing these is a pretty low bar. This [article](https://blog.sqreen.com/static-websites-security/) covers all these security issues well and includes solutions and deeper discussion.

Reviewer: We should name the link text as close to the title/link

```suggestion
Other possible attacks relate to using `iframes` or not removing referrer tags on external links. Thankfully, fixing these is a pretty low bar. The article [Security for Static Websites](https://blog.sqreen.com/static-websites-security/) covers all these security issues well and includes solutions and deeper discussion.
```

Reviewer 2: Good suggestion!

Reviewer 3: I agree, that would be a good change to make. Not seeing it here yet!

How do you secure secrets in the client-side world, if all the content and code is being shipped to the client browser? An answer can be found with the "[JAMstack](https://jamstack.org/)". Hopefully you are all fairly familiar with this concept given you are building "JAMstack" sites using Gatsby. For the uninitiated: you use HTML/CSS ([M]arkup) to build your site, [J]avaScript to make it dynamic, and [A]PIs (accessed by JavaScript) to provide content and features. All of this is done from the client-side.

The biggest security struggle with client-side sites is securing those APIs. Most APIs require an API key or another kind of authentication. Simply adding those secrets to the code only further obfuscates access to the same content. It does nothing to actually secure that content.

What's the solution? Well there are several problems to solve.

**Quick Definitions**:

> - Private - unique to a person or group of people using your site.

Reviewer: Use bold for definition like the other definition stylings in this file

```suggestion
- **Private** - unique to a person or group of people using your site.
```

> - Public - accessible to any and all users on your site.

Reviewer: ```suggestion
- **Public** - accessible to any and all users on your site.
```

> - Dynamic - any content that is updated more than once in a 5 minute interval and needs to be accessible to all users of the site.

Reviewer: ```suggestion
- **Dynamic** - any content that is updated more than once in a 5 minute interval and needs to be accessible to all users of the site.
```

> - Static - any content that changes less than once in a given 5 minute interval.

Reviewer: ```suggestion
- **Static** - any content that changes less than once in a given 5 minute interval.
```

Private content, whether dynamic or static, is another solution and is well documented in Gatsby's [authentication tutorial](/docs/authentication-tutorial/#security-notice).

**TL;DR:** Authenticate users using JSON Web Tokens(JWTs) and dynamically render pages only to authorized users. Any API calls that need to be made in order to fetch content can use the user's JWT and be verified by the API. If you need to access a third party API, any API Keys can be stored securely by your API which is securely authenticated to using the JWT.

This is pretty much the base of the "JAMstack" and allows for private content in client-side sites. Requiring user authentication will cover all use cases involving private data and will also help in situations where data is read publicly but written by authenticated users (e.g. CMS backend or comments section).

### Public Dynamic Content

This should be easy right? It is public content after all. But, if you need anonymous users to be able to both read and write this data, you should be concerned. Content read publicly, but written privately, was covered in the previous section. Admittedly, public reading and writing of content is less common.

> Medium's "clap" feature, up-voting on Hacker News or Reddit, liking something on Facebook or Twitter, these all require authentication. Even public APIs like GitHub's require authentication, even sites like Wikipedia; even if anyone in the world can create an account and access these tools.

Reviewer 3: ```suggestion
Medium's "clap" feature, up-voting on Hacker News or Reddit, liking something on Facebook or Twitter; these interactions all require authentication. Even public APIs like GitHub's require authentication, even sites like Wikipedia; even if anyone in the world can create an account and access these tools.
```

Assuming a nation state is not trying to attack you, what are you options for better security? You need an API.

One great option is to check out Netlify. They will provide great static hosting and allow you to deploy serverless functions right along side your Gatsby site. They are not the only ones: Azure, Google, Amazon, Zeit.co, and others all have API solutions.

What does this API allow? First, you can secure your database. Any API keys can be secured on your serverless...server 😭. The point is, they will be secure. But what does this gain you? After all, instead of pillaging your database directly you have given them an API through which to pillage.

> **Control:** You have gained the ability to add rate limiting, better logging, auto blocking of IPs, etc. You can throw the proverbial "secure API handbook" (except for the authentication part) at your users to make sure they behave. More aggressively, you could use a secure cookie for your site to uniquely and anonymously identify each user. This would assist in the previously mentioned security practices and, while not hard for a user to delete a cookie, it adds another layer of security.

Reviewer: 1) is there a link to a “Secure API Handbook”? I could not find a matching on google
2) i dont understand this sentence :( why i should throw a “handbook” at users? They need to read them? I expect more “at users login/authentification implementation”

Author: @muescha Okay, this is hilarious! It must just be an american idiom. to "Throw the book" at someone or something means to use everything in it. It came from the court system, [see dictionary.com](https://www.dictionary.com/browse/throw-the-book-at-someone).

the handbook doesn't exist...that's why it's "proverbial", meaning from a proverb, or "not real".

@marcysutton Thoughts on keeping or eliminating these american idioms?

Reviewer 3: @moonmeister if there's a way to write it that doesn't depend on slang or idioms, it would definitely make the text more inclusive!

Author: @marcysutton Yeah, I'll re-write it. Maybe see if I can find a good resource on API security even.

Reviewer 3: There is this cheat-sheet we could link to: https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/REST_Security_Cheat_Sheet.md

In the end if you need hardened security, you need authentication. If you want a simpler solution for basic features, these suggestions might just help. Remember, if you need access to a database and you already have authentication, do not ship those secure API keys in the client-side code!

### Exceptions

Every good rule has its exceptions. Never shipping API keys in client-side code is no exception. To be clear, you should never ship **SECURE** API keys (often called secrets) to a client. But, there is such a thing as non-secure/public API keys. These keys are generally used for identification and not for access control. Firebase authentication is a great example of this.

> Your site needs to be able to access your unique Firebase instance, to authenticate users. To do this, Firebase uses an API key to identify your app. This key is [designed to be public](https://stackoverflow.com/questions/37482366/is-it-safe-to-expose-firebase-apikey-to-the-public) and their documentation tells you to add it to your client-side code. [Firebase has other controls](https://stackoverflow.com/questions/35418143/how-to-restrict-firebase-data-modification) that determine what an anonymous user is allowed to do with this key.

Reviewer 3: ```suggestion
Your site needs to be able to access your unique Firebase instance to authenticate users. To do this, Firebase uses an API key to identify your app. This key is [designed to be public](https://stackoverflow.com/questions/37482366/is-it-safe-to-expose-firebase-apikey-to-the-public) and their documentation tells you to add it to your client-side code. [Firebase has other controls](https://stackoverflow.com/questions/35418143/how-to-restrict-firebase-data-modification) that determine what an anonymous user is allowed to do with this key.
```

This is also the case with Cloudinary. While they have a private API key (a secret) required to access private content, they also require an API key for [anonymous uploads](https://cloudinary.com/documentation/upload_images#unsigned_upload). This key is designed to be public and is used as a unique identifier of your cloud instance and not as a security measure.

It can difficult to know whether an API key should be kept secret. If the page giving you the keys does not specify, check the docs. If all else fails use a search engine: someone on Stack Overflow has likely already asked.

## Conclusion

Remember, keep it **secret**, keep it **safe**! Do not store API keys in your repository and do not ship them in your client-side code.

> Now go make awesome Gatsby sites that are completely secure! Fo more information on web security checkout the [OWASP Top Ten](https://www.owasp.org/images/7/72/OWASP_Top_10-2017_%28en%29.pdf.pdf). The [Open Web Application Security Project](https://www.owasp.org/index.php/Main_Page) is a excellent resource for security.

Reviewer: 1) We should mark the top ten Link that it is a PDF. 
2) better point to the project page with the PDF, because the PDF can be updated and unavailable. 

https://www.owasp.org/index.php/Category:OWASP_Top_Ten_Project

Author: I debated back and forth on this...you're probably correct.

---
title: Security for Modern Web Frameworks.

> date: 2019-03-27

Reviewer 3: @DSchau I merged this post without checking the date, do we need to do anything else for this one to be published? (does it need to go into scheduling?)

Author: It's already live on the site...it just shows as being backdated.

Reviewer 3: Well hey, that's fine by me if you're okay with it. :) congrats on publishing your post!
